test_reconnect_logic_stop_callback fails when 1.2.3.4/32 is not routable

Hi! We're seeing the `test_reconnect_logic_stop_callback` test fail in our build sandbox, that does not provide internet access.

```
____________________________________________________________________________ test_reconnect_logic_stop_callback _____________________________________________________________________________

patchable_api_client = <tests.conftest.patchable_api_client.<locals>.PatchableAPIClient object at 0x7f0e45b08360>

    @pytest.mark.asyncio
    async def test_reconnect_logic_stop_callback(patchable_api_client: APIClient):
        """Test that the stop_callback stops the ReconnectLogic."""
        cli = patchable_api_client
        rl = ReconnectLogic(
            client=cli,
            on_disconnect=AsyncMock(),
            on_connect=AsyncMock(),
            zeroconf_instance=get_mock_zeroconf(),
            name="mydevice",
        )
        await rl.start()
        assert rl._connection_state is ReconnectLogicState.DISCONNECTED
        await asyncio.sleep(0)
>       assert rl._connection_state is ReconnectLogicState.CONNECTING
E       assert <ReconnectLogicState.DISCONNECTED: 3> is <ReconnectLogicState.CONNECTING: 0>
E        +  where <ReconnectLogicState.DISCONNECTED: 3> = <aioesphomeapi.reconnect_logic.ReconnectLogic object at 0x7f0e45b21590>._connection_state
E        +  and   <ReconnectLogicState.CONNECTING: 0> = ReconnectLogicState.CONNECTING

tests/test_reconnect_logic.py:633: AssertionError
------------------------------------------------------------------------------------- Captured log call -------------------------------------------------------------------------------------
DEBUG    aioesphomeapi.reconnect_logic:reconnect_logic.py:290 Trying to connect to mydevice @ 1.2.3.4
DEBUG    aioesphomeapi.reconnect_logic:reconnect_logic.py:292 Connected lock acquired for mydevice @ 1.2.3.4
DEBUG    aioesphomeapi.connection:connection.py:331 mydevice @ 1.2.3.4: Connecting to IPv4Sockaddr(address='1.2.3.4', port=6052)
DEBUG    aioesphomeapi.connection:connection.py:278 Cleaning up connection to mydevice @ 1.2.3.4
WARNING  aioesphomeapi.reconnect_logic:reconnect_logic.py:173 Can't connect to ESPHome API for mydevice @ 1.2.3.4: Error connecting to [AddrInfo(family=<AddressFamily.AF_INET: 2>, type=<SocketKind.SOCK_STREAM: 1>, proto=6, sockaddr=IPv4Sockaddr(address='1.2.3.4', port=6052))]: [Errno 22] Invalid argument (SocketAPIError)
INFO     aioesphomeapi.reconnect_logic:reconnect_logic.py:303 Trying to connect to mydevice @ 1.2.3.4 in the background
DEBUG    aioesphomeapi.reconnect_logic:reconnect_logic.py:306 Retrying mydevice @ 1.2.3.4 in 2.00 seconds
DEBUG    aioesphomeapi.reconnect_logic:reconnect_logic.py:364 Starting zeroconf listener for mydevice
DEBUG    aioesphomeapi.reconnect_logic:reconnect_logic.py:232 Scheduling new connect attempt in 2.00 seconds
```

I can reproduce this outside the sandbox by adding a blackhole route towards the destination address.

```
ip route add blackhole 1.2.3.4/32
```

This will also happen when the machine has no route towards 1.2.3.4/32.

Would trying to connect to 127.0.0.1 be sufficient? The tests do work with that and it should always be available.

--- a/tests/conftest.py
+++ b/tests/conftest.py
@@ -57,7 +57,7 @@ def patchable_api_client() -> APIClient:
         pass
 
     cli = PatchableAPIClient(
-        address="1.2.3.4",
+        address="127.0.0.1",
         port=6052,
         password=None,
     )

--- a/tests/test_client.py
+++ b/tests/test_client.py
@@ -203,7 +203,7 @@ async def test_finish_connection_wraps_exceptions_as_unhandled_api_error(
 ) -> None:
     """Verify finish_connect re-wraps exceptions as UnhandledAPIError."""
 
-    cli = APIClient("1.2.3.4", 1234, None)
+    cli = APIClient("127.0.0.1", 1234, None)
     with patch("aioesphomeapi.client.APIConnection", PatchableAPIConnection):
         await cli.start_connection()
 
@@ -219,7 +219,7 @@ async def test_finish_connection_wraps_exceptions_as_unhandled_api_error(
 @pytest.mark.asyncio
 async def test_connection_released_if_connecting_is_cancelled() -> None:
     """Verify connection is unset if connecting is cancelled."""
-    cli = APIClient("1.2.3.4", 1234, None)
+    cli = APIClient("127.0.0.1", 1234, None)
     asyncio.get_event_loop()
 
     async def _start_connection_with_delay(*args, **kwargs):
@@ -930,7 +930,7 @@ async def test_noise_psk_handles_subclassed_string():
         pass
 
     cli = PatchableAPIClient(
-        address=Estr("1.2.3.4"),
+        address=Estr("127.0.0.1"),
         port=6052,
         password=None,
         noise_psk=Estr("QRTIErOb/fcE9Ukd/5qA3RGYMn0Y+p06U58SCtOXvPc="),
@@ -966,7 +966,7 @@ async def test_noise_psk_handles_subclassed_string():
 async def test_no_noise_psk():
     """Test not using a noise_psk."""
     cli = APIClient(
-        address=Estr("1.2.3.4"),
+        address=Estr("127.0.0.1"),
         port=6052,
         password=None,
         noise_psk=None,
@@ -982,7 +982,7 @@ async def test_no_noise_psk():
 async def test_empty_noise_psk_or_expected_name():
     """Test an empty noise_psk is treated as None."""
     cli = APIClient(
-        address=Estr("1.2.3.4"),
+        address=Estr("127.0.0.1"),
         port=6052,
         password=None,
         noise_psk="",

--- a/tests/test_log_runner.py
+++ b/tests/test_log_runner.py
@@ -46,7 +46,7 @@ async def test_log_runner(
     async_zeroconf = get_mock_async_zeroconf()
 
     cli = PatchableAPIClient(
-        address=Estr("1.2.3.4"),
+        address=Estr("127.0.0.1"),
         port=6052,
         password=None,
         noise_psk=None,
@@ -115,7 +115,7 @@ async def test_log_runner_reconnects_on_disconnect(
     async_zeroconf = get_mock_async_zeroconf()
 
     cli = PatchableAPIClient(
-        address=Estr("1.2.3.4"),
+        address=Estr("127.0.0.1"),
         port=6052,
         password=None,
         noise_psk=None,
@@ -195,7 +195,7 @@ async def test_log_runner_reconnects_on_subscribe_failure(
     async_zeroconf = get_mock_async_zeroconf()
 
     cli = PatchableAPIClient(
-        address=Estr("1.2.3.4"),
+        address=Estr("127.0.0.1"),
         port=6052,
         password=None,
         noise_psk=None,

--- a/tests/test_reconnect_logic.py
+++ b/tests/test_reconnect_logic.py
@@ -102,7 +102,7 @@ async def test_reconnect_logic_name_from_host_and_set():
 async def test_reconnect_logic_name_from_address():
     """Test that the name is set correctly from the address."""
     cli = APIClient(
-        address="1.2.3.4",
+        address="127.0.0.1",
         port=6052,
         password=None,
     )
@@ -119,14 +119,14 @@ async def test_reconnect_logic_name_from_address():
         on_connect=on_connect,
         zeroconf_instance=get_mock_zeroconf(),
     )
-    assert cli.log_name == "1.2.3.4"
+    assert cli.log_name == "127.0.0.1"
 
 
 @pytest.mark.asyncio
 async def test_reconnect_logic_name_from_name():
     """Test that the name is set correctly from the address."""
     cli = APIClient(
-        address="1.2.3.4",
+        address="127.0.0.1",
         port=6052,
         password=None,
     )
@@ -144,7 +144,7 @@ async def test_reconnect_logic_name_from_name():
         zeroconf_instance=get_mock_zeroconf(),
         name="mydevice",
     )
-    assert cli.log_name == "mydevice @ 1.2.3.4"
+    assert cli.log_name == "mydevice @ 127.0.0.1"
 
 
 @pytest.mark.asyncio
@@ -201,7 +201,7 @@ async def test_reconnect_logic_state(patchable_api_client: APIClient):
         name="mydevice",
         on_connect_error=on_connect_fail,
     )
-    assert cli.log_name == "mydevice @ 1.2.3.4"
+    assert cli.log_name == "mydevice @ 127.0.0.1"
 
     with patch.object(cli, "start_connection", side_effect=APIConnectionError):
         await rl.start()
@@ -274,7 +274,7 @@ async def test_reconnect_retry(
         name="mydevice",
         on_connect_error=on_connect_fail,
     )
-    assert cli.log_name == "mydevice @ 1.2.3.4"
+    assert cli.log_name == "mydevice @ 127.0.0.1"
     caplog.clear()
 
     with patch.object(cli, "start_connection", side_effect=APIConnectionError):
@@ -288,9 +288,9 @@ async def test_reconnect_retry(
     assert len(on_connect_fail_called) == 1
     assert isinstance(on_connect_fail_called[-1], APIConnectionError)
     assert rl._connection_state is ReconnectLogicState.DISCONNECTED
-    assert "connect to ESPHome API for mydevice @ 1.2.3.4" in caplog.text
+    assert "connect to ESPHome API for mydevice @ 127.0.0.1" in caplog.text
     for record in caplog.records:
-        if "connect to ESPHome API for mydevice @ 1.2.3.4" in record.message:
+        if "connect to ESPHome API for mydevice @ 127.0.0.1" in record.message:
             assert record.levelno == logging.WARNING
 
     caplog.clear()
@@ -307,9 +307,9 @@ async def test_reconnect_retry(
     assert len(on_connect_fail_called) == 2
     assert isinstance(on_connect_fail_called[-1], APIConnectionError)
     assert rl._connection_state is ReconnectLogicState.DISCONNECTED
-    assert "connect to ESPHome API for mydevice @ 1.2.3.4" in caplog.text
+    assert "connect to ESPHome API for mydevice @ 127.0.0.1" in caplog.text
     for record in caplog.records:
-        if "connect to ESPHome API for mydevice @ 1.2.3.4" in record.message:
+        if "connect to ESPHome API for mydevice @ 127.0.0.1" in record.message:
             assert record.levelno == logging.DEBUG
 
     caplog.clear()
@@ -320,7 +320,7 @@ async def test_reconnect_retry(
         await asyncio.sleep(0)
         await asyncio.sleep(0)
 
-    assert "connect to ESPHome API for mydevice @ 1.2.3.4" not in caplog.text
+    assert "connect to ESPHome API for mydevice @ 127.0.0.1" not in caplog.text
     assert len(on_disconnect_called) == 0
     assert len(on_connect_called) == 1
     assert len(on_connect_fail_called) == 2
@@ -372,7 +372,7 @@ DNS_POINTER = DNSPointer(
                 _TYPE_A,
                 _CLASS_IN,
                 1000,
-                ip_address("1.2.3.4").packed,
+                ip_address("127.0.0.1").packed,
             ),
             True,
             ReconnectLogicState.READY,
@@ -403,7 +403,7 @@ async def test_reconnect_zeroconf(
         name="mydevice",
         on_connect_error=AsyncMock(),
     )
-    assert cli.log_name == "mydevice @ 1.2.3.4"
+    assert cli.log_name == "mydevice @ 127.0.0.1"
 
     with patch.object(
         cli, "start_connection", side_effect=quick_connect_fail
@@ -474,7 +474,7 @@ async def test_reconnect_zeroconf_not_while_handshaking(
         name="mydevice",
         on_connect_error=AsyncMock(),
     )
-    assert cli.log_name == "mydevice @ 1.2.3.4"
+    assert cli.log_name == "mydevice @ 127.0.0.1"
 
     with patch.object(
         cli, "start_connection", side_effect=quick_connect_fail
@@ -531,7 +531,7 @@ async def test_connect_task_not_cancelled_while_handshaking(
         name="mydevice",
         on_connect_error=AsyncMock(),
     )
-    assert cli.log_name == "mydevice @ 1.2.3.4"
+    assert cli.log_name == "mydevice @ 127.0.0.1"
 
     with patch.object(
         cli, "start_connection", side_effect=quick_connect_fail
@@ -591,7 +591,7 @@ async def test_connect_aborts_if_stopped(
         name="mydevice",
         on_connect_error=AsyncMock(),
     )
-    assert cli.log_name == "mydevice @ 1.2.3.4"
+    assert cli.log_name == "mydevice @ 127.0.0.1"
 
     with patch.object(
         cli, "start_connection", side_effect=quick_connect_fail
@@ -694,7 +694,7 @@ async def test_handling_unexpected_disconnect(aiohappyeyeballs_start_connection)
     async_zeroconf = get_mock_async_zeroconf()
 
     cli = PatchableAPIClient(
-        address="1.2.3.4",
+        address="127.0.0.1",
         port=6052,
         password=None,
         noise_psk=None,
@@ -770,7 +770,7 @@ async def test_backoff_on_encryption_error(
     async_zeroconf = get_mock_async_zeroconf()
 
     cli = PatchableAPIClient(
-        address="1.2.3.4",
+        address="127.0.0.1",
         port=6052,
         password=None,
         noise_psk="",

